main/src/ca/uwaterloo/flix/language/phase/Stratifier.scala

```diff
@@ -37,13 +38,125 @@ object Stratifier extends Phase[TypedAst.Root, TypedAst.Root] {
+      edges = edges.get(from) match {
+        case Some(m) => edges + (from -> (m + (to -> weight)))
+        case None => edges + (from -> Map(to -> weight))
+      }
+      vertices += from
+      vertices += to
+      this
+    }
+
+    /**
+      * Get the list of edges that a vertex has edges to
+      */
+    def lookup(from: Symbol.TableSym): Option[Map[Symbol.TableSym, Int]] = {
+      edges.get(from)
+    }
+  }
+
+  /**
+    * Use the contraints to create an adjacency matrix
+    *
+    * A constraint of:
+    *   P :- Q creates an edge Q -> P of weight 0
+    *   P :- !Q creates an edge Q -> P of weight -1
+    */
+  def createGraph(constraints: List[TypedAst.Constraint]): Graph = {
+    constraints.foldRight(new Graph)((constraint: TypedAst.Constraint, graph: Graph) => constraint.head match {
+      case TypedAst.Predicate.Head.Positive(headSym, _, _) =>
+        if (graph.head == null ) {
+          graph.head = headSym
+        }
+        constraint.body.foldRight(graph)((pred: TypedAst.Predicate.Body, graph: Graph) => pred match {
+          case TypedAst.Predicate.Body.Negative(predSym, _, _) =>
+            graph.insert(headSym, predSym, -1)
+          case TypedAst.Predicate.Body.Positive(predSym, _, _) =>
+            graph.insert(headSym, predSym, 0)
+          case _:TypedAst.Predicate.Body.Filter => graph
+          case _:TypedAst.Predicate.Body.Loop => graph
+        })
+      case TypedAst.Predicate.Head.Negative(headSym, _, _) => ??? // Can't have negative heads
+      case TypedAst.Predicate.Head.True(_) => graph
+      case TypedAst.Predicate.Head.False(_) => graph
+    })
+  }
+
+  /**
+    * Returns true iff a graph has a negative cycle
+    */
+  def hasNegativeCycles(g: Graph): Boolean = {
+    // The Bellman Ford algorithm, we can use it to find negative cycles
+
+    // Initialize the distance and predecessor arrays to default values
+    var distPred: Map[Symbol.TableSym, (Int, Symbol.TableSym)] = g.vertices.toList.zip(List.fill(g.vertices.size)((1, null))).toMap
+    // The source has 0 weight
+    distPred = distPred.updated(g.head, (0, null))
+
+    // Repeatedly relax the edges
+    g.vertices.foreach(_ => {
+      g.edges.foreach((fromEntry: (Symbol.TableSym, Map[Symbol.TableSym, Int])) => {
+        fromEntry._2.foreach(toEntry => {
+          val from = fromEntry._1
+          val to = toEntry._1
+          val weight = toEntry._2
+          // Relax the edge from-to
+          if (distPred.get(from).get._1 + weight < distPred.get(to).get._1) {
+            distPred = distPred.updated(to, (distPred.get(from).get._1 + weight, from))
+          }
+        })
+      })
+    })
+
+    // At this point, we must have found the shortest path, so if there is a
+    // shorter path, then there is a negative cycle
+    g.edges.foreach((fromEntry: (Symbol.TableSym, Map[Symbol.TableSym, Int])) => {
+      fromEntry._2.foreach(toEntry => {
```

Comment:
Here you can use the pattern matching for loop of Scala. Something like:

`for ((x, y) <- ...)`